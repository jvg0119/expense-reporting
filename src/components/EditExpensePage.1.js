import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense, removeExpense } from '../actions/expenses';

// const EditExpensePage = (props) => {
export const EditExpensePage = (props) => { 
  return (
    <div>
      <ExpenseForm
        expense={props.expense}
        onSubmit={(expense) => {
          // props.dispatch(editExpense(props.expense.id, expense));
          props.editExpense(props.expense.id, expense);
          props.history.push('/');
        }}
      />
      <button onClick={() => {
        // props.dispatch(removeExpense({ id: props.expense.id }));
        props.removeExpense({ id: props.expense.id });
        props.history.push('/');
      }}>Remove</button>
    </div>
  );
};

const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find((expense) => expense.id === props.match.params.id)
  };
};

const mapDispatchToProps = (dispatch, props) => {
  return {
    editExpense: (id, expense) => {
      dispatch(editExpense(id, expense))
    },
    removeExpense: (id) => {
      dispatch(removeExpense(id))
    }
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);

// Experiment     --- this works   no test for this experiment 
// using the mapDispatchToProps  but keeping the stateless component 
// not using the class component 
