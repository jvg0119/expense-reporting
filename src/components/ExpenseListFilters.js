import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from '../actions/filters';

export class ExpenseListFilters extends React.Component {
  
  state = {
    calendarFocused: null
  };

  onDatesChange = ({ startDate, endDate }) => {
    // this.props.dispatch(setStartDate(startDate));
    // this.props.dispatch(setEndDate(endDate));
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };

  onFocusChange = (calendarFocused) => {
    // console.log('from onFocusChange >>> ', calendarFocused)
    // below are different versions of setState 
    // this.setState({ calendarFocused })
    // this.setState({ calendarFocused: calendarFocused })
    // this.setState(() => {
    //   return {
    //     calendarFocused: calendarFocused
    //   }
    // })
    // this.setState(() => ({ calendarFocused: calendarFocused }) );
    this.setState(() => ({ calendarFocused }));
  }
  
  onTextChange = (e) => {
    // this.props.dispatch(setTextFilter(e.target.value));
    this.props.setTextFilter(e.target.value);
    // console.log('onTextChange is running !!!')
  }

  onSortBy = (e) => {
    if (e.target.value === 'date') {
      // this.props.dispatch(sortByDate());
      this.props.sortByDate();
    } else if (e.target.value === 'amount') {
      this.props.sortByAmount();
    }
  }

  render() {
    return (
      <div>
        <input
          type="text"
          value={this.props.filters.text}
          // onChange={(e) => {
          //   this.props.dispatch(setTextFilter(e.target.value));
          // }}
          onChange={this.onTextChange}
        />
        <select
          value={this.props.filters.sortBy}
          // onChange={(e) => {
          //   if (e.target.value === 'date') {
          //     this.props.dispatch(sortByDate());
          //   } else if (e.target.value === 'amount') {
          //     this.props.dispatch(sortByAmount());
          //   }
          // }}
          onChange={this.onSortBy}
        >
          <option value="date">Date</option>
          <option value="amount">Amount</option>
        </select>
        <DateRangePicker
          startDate={this.props.filters.startDate}
          endDate={this.props.filters.endDate}
          onDatesChange={this.onDatesChange}
          focusedInput={this.state.calendarFocused}
          onFocusChange={this.onFocusChange}
          showClearDates={true}
          numberOfMonths={1}
          isOutsideRange={() => false}
        />
      </div>
    );
  }
};

const mapStateToProps = (state) => {
  return {
    filters: state.filters
  };
};

const mapDispatchToProps = (dispatch) => ({
  setTextFilter: (text) => dispatch(setTextFilter(text)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount()),
  setStartDate: (startDate) => dispatch(setStartDate(startDate)),
  setEndDate: (endDate) => dispatch(setEndDate(endDate))
})

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);

// refactor these from    ExpenseListFilters.js 
//   break out the inline functions 
//   add mapDispatchToProps


