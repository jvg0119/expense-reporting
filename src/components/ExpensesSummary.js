import React from 'react'; 
import selectExpenses from '../selectors/expenses';
import selectExpensesTotal from '../selectors/expenses-total';
import { connect } from 'react-redux';
import numeral from 'numeral';

export const ExpensesSummary = ({expenseCount, expensesTotal}) => {
  const expenseWord = expenseCount === 1 ? 'expense' : 'expenses';
  const formattedExpensesTotal = numeral(expensesTotal/100).format('$0,0.00'); 
  return (
    <div>
      Viewing {expenseCount} {expenseWord} totalling {formattedExpensesTotal}
    </div>
  )
}  

const mapStateToProps = (state) => {
  const visibleExpenses = selectExpenses(state.expenses, state.filters);
  return {
    expenseCount: visibleExpenses.length,
    expensesTotal: selectExpensesTotal(visibleExpenses)
  }
}


// below is alt method but not setup for testing 
// const ExpensesSummary = (props) => {
//   const expenseWord = props.expenseCount.length === 1 ? "expense" : "expenses";
//   const formattedExpensesTotal = numeral(props.expensesTotal/100).format('$0,0.00')
//   return (
//     <div>
//       Viewing {props.expenseCount.length} {expenseWord} totalling {formattedExpensesTotal}
//     </div>
//   )
// }

// const mapStateToProps = (state, props) => {
//   const visibleExpenses = selectExpenses(state.expenses, state.filters)
//   return {
//     expenseCount: visibleExpenses,
//     expensesTotal: selectExpensesTotal(visibleExpenses)
//   }
// }
  
export default connect(mapStateToProps)(ExpensesSummary);
