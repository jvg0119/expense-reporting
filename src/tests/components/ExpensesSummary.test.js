import React from 'react';
import { ExpensesSummary } from '../../components/ExpensesSummary';
import { shallow } from 'enzyme';

test('should render expense summary with 1 expense', () => {
  const wrapper = shallow(<ExpensesSummary expenseCount={1} expensesTotal={100}/>);
  expect(wrapper).toMatchSnapshot();
})

test('should render expense summary with multiple expenses', () => {
  const wrapper = shallow(<ExpensesSummary expenseCount={5} expensesTotal={10000}/>);
  expect(wrapper).toMatchSnapshot();
})
