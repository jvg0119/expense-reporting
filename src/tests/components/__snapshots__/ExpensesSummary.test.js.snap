// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render expense summary with 1 expense 1`] = `
<div>
  Viewing 
  1
   
  expense
   totalling 
  $1.00
</div>
`;

exports[`should render expense summary with multiple expenses 1`] = `
<div>
  Viewing 
  5
   
  expenses
   totalling 
  $100.00
</div>
`;
